kong/plugins/request-transformer/access.lua

```diff
@@ -135,106 +150,163 @@ local function transform_querystrings(conf)
           parameters:set_simple(name, value)
         end
       end
-      local new_data = parameters:tostring()
-      req_set_header(CONTENT_LENGTH, string_len(new_data))
-      req_set_body_data(new_data)
+      new_body = parameters:tostring()
+    elseif content_type == JSON then
+      local body = req_get_body_data()
+      local parameters = parse_json(req_get_body_data())
+      if parameters ~= nil then
+        for _, name, value in iter(conf.replace.body) do
+          if parameters[name] then
+            parameters[name] = value
+          end
+        end
+        new_body = cjson.encode(parameters)
+      else
+        new_body = body
+      end    
     end
+    req_set_header(CONTENT_LENGTH, string_len(new_body))
+    req_set_body_data(new_body)
   end
 
-  -- Add form parameter(s)
-  if conf.add.form then
-    local content_type = get_content_type()
-    if content_type and stringy.startswith(content_type, FORM_URLENCODED) then
-      -- Call req_read_body to read the request body first
-      req_read_body()
-
+  -- Add parameters
+  if conf.add.body then
+    
+    -- Call req_read_body to read the request body first
+    req_read_body()
+    local new_body =  "" 
+    if content_type == ENCODED then
       local parameters = req_get_post_args()
-      for _, name, value in iter(conf.add.form) do
+      for _, name, value in iter(conf.add.body) do
         if not parameters[name] then
           parameters[name] = value
         end
       end
-      local encoded_args = encode_args(parameters)
-      req_set_header(CONTENT_LENGTH, string_len(encoded_args))
-      req_set_body_data(encoded_args)
-    elseif content_type and stringy.startswith(content_type, MULTIPART_DATA) then
-      -- Call req_read_body to read the request body first
-      req_read_body()
-
+      new_body = encode_args(parameters)
+    elseif content_type == MULTI then
       local body = req_get_body_data()
-      local parameters = multipart(body and body or "", content_type)
-      for _, name, value in iter(conf.add.form) do
+      local parameters = multipart(body and body or "", content_type_value)
+      for _, name, value in iter(conf.add.body) do
         if not parameters:get(name) then
           parameters:set_simple(name, value)
         end
       end
-      local new_data = parameters:tostring()
-      req_set_header(CONTENT_LENGTH, string_len(new_data))
-      req_set_body_data(new_data)
+      new_body = parameters:tostring()
+    elseif content_type == JSON then
+      local body = req_get_body_data()
+      local parameters = parse_json(req_get_body_data())
+      if parameters ~= nil then
+        for _, name, value in iter(conf.add.body) do
+          if not parameters[name] then
+            parameters[name] = value
+          end
+        end
+        new_body = cjson.encode(parameters)
+      else
+        new_body = body
+      end    
+    end
+    req_set_header(CONTENT_LENGTH, string_len(new_body))
```

Comment:
Couldn't the content length be set only once at the end of this function instead of every time the body gets modified?